httpclient_android/http_client/src/main/java/com/backblaze/b2/client/webapiurlconnection/android/UrlConnection/B2WebApiHttpClientImpl.java

```diff
@@ -0,0 +1,398 @@
+                final StringBuilder response = new StringBuilder();
+                String line;
+                while ((line = input.hasNext())) {
+                    response.append(line);
+                }
+                input.close();
+                throw extractExceptionFromErrorResponse(client, response.toString());
+            }
+        } catch (IOException e) {
+            throw translateToB2Exception(e, clientUrl);
+        }
+    }
+
+    /**
+     * HEADSs to a web service that returns content, and returns the headers.
+     *
+     * @param url the url to head to
+     * @param headersOrNull the headers, if any.
+     * @return the headers of the response.
+     * @throws B2Exception if there's any trouble
+     */
+    @Override
+    public B2Headers head(String url, B2Headers headersOrNull)
+            throws B2Exception {
+        try {
+            final URL clientUrl = new URL(url);
+            URLConnection head;
+            if (clientUrl.getProtocol().equals("https")) {
+                 head = (HttpsURLConnection) clientUrl.openConnection();
+            } else {
+                head = (HttpURLConnection) clientUrl.openConnection();
+            }
+            if (headersOrNull != null) {
+                makeHeaders(headersOrNull).forEach((name, value) -> head.setRequestProperty(name, value));
+            }
+            head.setDoOutput(true);
+            int statusCode = head.getResponseCode();
+            if (statusCode >= 200 && statusCode < 300) {
+                final B2HeadersImpl.Builder builder = B2HeadersImpl.builder();
+                Arrays.stream(head.getHeaderFields().entrySet().forEach(header -> builder.set(header.getName(), header.getValue())));
+                return builder.build();
+            } else {
+                throw B2Exception.create(null, statusCode, null, "");
+            }
+        } catch (IOException e) {
+            throw translateToB2Exception(e, url);
+        }
+        finally {
+            closeQuietly(response);
+        }
+    }
+
+    @Override
+    public void close() {
+        clientFactory.close();
+    }
+
+    private String postJsonAndReturnString(String url,
+                                           B2Headers headersOrNull,
+                                           Object request) throws B2Exception {
+        String requestString = parseToStringUsingBzJson(request);
+        return postAndReturnString(url, headersOrNull, requestString);
+    }
+
+    /**
+     * POSTs to a web service that returns content, and returns the content
+     * as a single string.
+     *
+     * @param url the url to post to
+     * @param headersOrNull the headers, if any.
+     * @param requestString the string to post.
+     * @return the body of the response.
+     * @throws B2Exception if there's any trouble
+     */
+    private String postAndReturnString(String url, B2Headers headersOrNull, String requestString)
```

Comment:
can `requestString` be wrapped as an InputStream and call the other overload of `postAndReturnString` so that there's only one core implementation?